### PR TITLE
Request Missing Parents Recursively via P2P

### LINKED ISSUE #1889: Request Block Parents from Peer Recursively If Missing When Processing
Currently we run into an issue when syncing when if a node does not have a parent block when processing, it will fail and its chain will never advance. Instead, we should request parent blocks from peers. @nisdas outlined the following as a solution:

- Request for block by hash, 
- Once we receive it we check again for its parent, we can recursively do this until maybe a limit of 10 blocks or something.
- If we have like 10 blocks missing that means something is deeply wrong,
- Have an in-memory map of blocks similar to initial-sync
- We can use that to store any blocks that cannot be processed yet 
- Add in a new field in the regular sync struct, currentHead 
- using this we can pre-check received blocks, if the block is more than 1 slot ahead then we keep it in the map, we can keep requesting for parent blocks until we get one that has is 1 slot ahead from the currentHead

### PR DESCRIPTION
**Pull request by rauljordan**
This resolves #1889 

---

# Description

**Write why you are making the changes in this pull request**

If we are missing a block parent, we need to recursively request its parent until we are no longer missing them in our persistent storage.

**Write a summary of the changes you are making**

This PR implements the suggested logic from #1889.

### CODE REVIEW

File: beacon-chain/sync/receive_block.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	rs.blocksAwaitingProcessingLock.Lock()
+	defer rs.blocksAwaitingProcessingLock.Unlock()
+	delete(rs.blocksAwaitingProcessing, blockRoot)
+	blocksAwaitingProcessingGauge.Dec()
+}
+
+func (rs *RegularSync) hasChild(blockRoot [32]byte) (p2p.Message, bool) {
+	rs.blocksAwaitingProcessingLock.Lock()
+	defer rs.blocksAwaitingProcessingLock.Unlock()
+	child, ok := rs.blocksAwaitingProcessing[blockRoot]
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]:
```go
return rs.blocksAwaitingProcessing[blockRoot]
```

Comment (rauljordan) [reply beacon-chain/sync/receive_block.go]: Maps are weird in the sense that if we do that, it only returns the child and not the bool ok value. We have to unpack it in order to extract both @prestonvanloon.

File: beacon-chain/sync/receive_block.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	}
+
+	// We check if we have the block's parents saved locally.
+	parentRoot := bytesutil.ToBytes32(block.ParentRootHash32)
+	hasParent := rs.db.HasBlock(parentRoot)
+	span.AddAttributes(trace.BoolAttribute("hasParent", hasParent))
+
+	if !hasParent {
+		// If we do not have the parent, we insert it into a pending block's map.
+		rs.insertPendingBlock(ctx, parentRoot, msg)
```
Comment (prestonvanloon) [inline beacon-chain/sync/receive_block.go]: I think you should also broadcast a request for the parent block. Otherwise we may never receive it

Comment (prestonvanloon) [reply beacon-chain/sync/receive_block.go]: nvm i see you do this

File: beacon-chain/sync/receive_block.go (hunk 1 above)
Comment (prestonvanloon) [reply beacon-chain/sync/receive_block.go]: got it

**Pull request merged**